contrib/go/src/python/pants/contrib/go/tasks/go_compile.py

```diff
@@ -49,7 +51,8 @@ def execute(self):
           self._go_install(vt.target, gopath)
         if self.is_binary(vt.target):
           binary_path = os.path.join(gopath, 'bin', os.path.basename(vt.target.address.spec_path))
-          self.context.products.get_data('exec_binary')[vt.target] = binary_path
+          go_exec_binary[vt.target] = binary_path
+          go_deployable_archive.add(vt.target, os.path.dirname(binary_path)).append(os.path.basename(binary_path))
```

Comment:
This pattern (explicitly calling basename) seems less error prone than what you're doing for `cpp`, which could likely be missed in a refactor.